intermediate_source/transformer_building_blocks.py

```diff
@@ -0,0 +1,686 @@
+# NJT also composes with the ``FlexAttention`` module. This is a generalization
+# of the ``MultiheadAttention`` layer that allows for arbitrary modifications
+# to the attention score. The example below takes the ``alibi_mod``
+# that implements `ALiBi <https://arxiv.org/abs/2108.12409>`_ from
+# `attention gym <https://github.com/pytorch-labs/attention-gym>`_ and uses it
+# with nested input tensors.
+
+from torch.nn.attention.flex_attention import flex_attention
+
+def generate_alibi_bias(H: int):
+    """Returns an alibi bias score_mod given the number of heads H
+    Args:
+        H: number of heads
+    Returns:
+        alibi_bias: alibi bias score_mod
+    """
+    def alibi_mod(score, b, h, q_idx, kv_idx):
+        scale = torch.exp2(-((h + 1) * 8.0 / H))
+        bias = (q_idx - kv_idx) * scale
+        return score + bias
+    return alibi_mod
+
+query, key, value, _ = gen_batch(N, E_q, E_k, E_v, device)
+n_heads, D = 8, E_q // 8
+alibi_score_mod = generate_alibi_bias(n_heads)
+query = (
+    query.unflatten(-1, [n_heads, D]).transpose(1, 2).detach().requires_grad_()
+)
+key = key.unflatten(-1, [n_heads, D]).transpose(1, 2).detach().requires_grad_()
+value = (
+    value.unflatten(-1, [n_heads, D]).transpose(1, 2).detach().requires_grad_()
+)
+out_flex2 = flex_attention(query, key, value, score_mod=alibi_score_mod)
+
+###############################################################################
+# Packed Projection
+# -----------------
+# 
+# Packed projection is a technique that makes use of the fact that when the input
+# for projection (matrix multiplications) are the same (self-attention), we can pack the projection
+# weights and biases into single tensors. It is especially useful when the individual
+# projections are memory bound rather than compute bound. There are
+# two examples that we will demonstrate here:
+# 
+# * Input projection for MultiheadAttention
+# * SwiGLU activation in feed-forward network of Transformer Layer
+# 
+# Input projection for MultiheadAttention
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Recall that when doing self-attention, the ``query``, ``key`` and ``value``
+# are the same tensor. Each of these tensors is projected with a 
+# ``Linear(E_q, E_total)`` layer. Instead, we can pack this into one layer,
+# which is what we do in the MultiheadAttention layer above.
+# 
+# Let us compare the performance of the packed projection against the usual method:
+
+class InputProjection(nn.Module):
+    def __init__(self, E_q, E_total, bias=False, device=None, dtype=None):
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        super().__init__()
+        self.q_proj = nn.Linear(E_q, E_total, bias=bias, **factory_kwargs)
+        self.k_proj = nn.Linear(E_q, E_total, bias=bias, **factory_kwargs)
+        self.v_proj = nn.Linear(E_q, E_total, bias=bias, **factory_kwargs)
+
+    def forward(self, query):  
+        return self.q_proj(query), self.k_proj(query), self.v_proj(query)
```

Comment:
nit: being pedantic here, but maybe `x` or something is a better name than `query` to avoid wrongly indicating that this is query-specific